clippy_lints/src/string_patterns.rs

```diff
@@ -0,0 +1,223 @@
+            SINGLE_CHAR_PATTERN,
+            arg.span,
+            "single-character string constant used as pattern",
+            "consider using a `char`",
+            hint,
+            *applicability,
+        );
+    }
+}
+
+fn check_manual_pattern_char_comparison(cx: &LateContext<'_>, arg: &Expr<'_>, applicability: Applicability) {
+    if let ExprKind::Closure(closure) = arg.kind
+        && let body = cx.tcx.hir().body(closure.body)
+        && let Some(param_name) = body.params[0].pat.simple_ident()
+    {
+        let ex = body.value;
+        let mut set_chars: Vec<String> = Vec::new();
+
+        // We want to retrieve all the comparisons done.
+        // They are ordered in a nested way and so we need to traverse the AST to collect them all.
+        for_each_expr(ex, |sub_expr| -> ControlFlow<(), Descend> {
+            match sub_expr.kind {
+                ExprKind::Binary(op, left, right) => {
+                    // Avoid matching code that contains complex comparisons such as functions calls etc..
+                    // (ex: c.is_whitespace())
+                    if matches!(
+                        left.kind,
+                        ExprKind::Binary(_, _, _) | ExprKind::Match(_, _, _) | ExprKind::Lit(_) | ExprKind::Path(_)
+                    ) && matches!(
+                        right.kind,
+                        ExprKind::Binary(_, _, _) | ExprKind::Match(_, _, _) | ExprKind::Lit(_) | ExprKind::Path(_)
+                    ) {
+                        if matches!(op.node, BinOpKind::Eq | BinOpKind::Or) {
+                            ControlFlow::Continue(Descend::Yes)
+                        } else {
+                            ControlFlow::Continue(Descend::No)
+                        }
+                    } else {
+                        ControlFlow::Break(())
+                    }
+                },
+                ExprKind::Match(match_value, arms, _) => {
+                    if snippet(cx, match_value.span, "") != param_name.name.as_str() || arms.len() != 2 {
+                        return ControlFlow::Break(());
+                    }
+                    if let PatKind::Or(pats) = arms[0].pat.kind {
+                        for pat in pats {
+                            if let PatKind::Lit(expr) = pat.kind
+                                && let ExprKind::Lit(lit) = expr.kind
+                                && let LitKind::Char(c) = lit.node
+                            {
+                                set_chars.push(format!("'{}'", c.escape_default()));
+                            }
+                        }
+                    } else if let PatKind::Lit(expr) = arms[0].pat.kind
+                        && let ExprKind::Lit(lit) = expr.kind
+                        && let LitKind::Char(c) = lit.node
+                    {
+                        set_chars.push(format!("'{}'", c.escape_default()));
+                    }
+                    ControlFlow::Continue(Descend::No)
```

Comment:
If we see a pattern that isn't a literal we need to avoid linting also, e.g.

```rust
"".find(|c| matches!(c, 'a' | '1'..'4'));
"".find(|c| c == 'a' || matches!(c, '1'..'4'));
```

`walk_short` can make it a little nicer by combining the `Or`/`Lit` cases, if it returns `false` we'd need to return `Break`

```rust
arm.pat.walk_short(|pat| match pat.kind {
    PatKind::Lit(expr) if let ExprKind::Lit.. => { /* push */ }
    PatKind::Or(_) => true,
    _ => false,
});
```

Reply:
`walk_short` makes it so clearer. Thank you !